M2/Macaulay2/packages/Style.m2

```diff
@@ -3,16 +3,184 @@ newPackage( "Style",
      AuxiliaryFiles => true,
      Headline => "style sheets and images for the documentation",
      Keywords => {"Documentation"},
-     Version => "1.0"
+     Version => "1.1"
      )
 
+---------------------------------------------
+-- grammar generation code adapted from    --
+-- M2/Macaulay2/editors/make-M2-symbols.m2 --
+---------------------------------------------
+
+export {"generateGrammar"}
+
+importFrom(Core, "sortBy")
+
+-- TODO: Move these two elsewhere:
+Function and Function := (f, g) -> s -> f s and g s
+Function or  Function := (f, g) -> s -> f s or  g s
+
+is := X -> (name, symb) -> instance(value symb, X)
+
+isAlpha        := s -> match("^[[:alpha:]]+$", s)
+isAlphaNumeric := s -> match("^[[:alnum:]]+$", s)
+
+-- Things we want to highlight:
+isType     := is Type
+isKeyword  := is Keyword
+isFunction := is Function
+isConst    := (name, symb) -> (isAlphaNumeric name
+    and not (isFunction or isType or isKeyword) (name, symb)
+    and (symb === symbol null or value symb =!= null))
+
+okay := method()
+okay(String, Keyword) :=
+okay(String, Symbol)  := (name, pkg) -> length name > 1 and isAlphaNumeric name
+
+cachedSymbols = new CacheTable
+STRINGS = format "///\\\\(/?/?[^/]\\\\|\\\\(//\\\\)*////[^/]\\\\)*\\\\(//\\\\)*///"
+
+-- This banner is added to the top of generated grammars
+banner := "Auto-generated for Macaulay2-@M2VERSION@. Do not modify this file manually."
+
+generateSymbols = () -> (
+    ----------------------------------------------------------------------------
+    -- Get a list of all symbols visible just after loading preloaded packages
+    allPkgNames := separate(" ", version#"packages") | {"Core"};
+    loadedPkgNames := join(Core#"preloaded packages",
+	{"Core", "Text", "Parsing", "SimpleDoc"});
+    symbols := unique (sortBy ascii @@ first) join(
+	apply(allPkgNames, pkgname -> (pkgname, symbol Core)),
+	flatten apply(loadedPkgNames, pkgname -> (
+		pkg := needsPackage pkgname;
+		select(pairs pkg.Dictionary, okay))));
+
+    if length symbols < 1500 then error "expected more entries for M2-symbols";
+
+    -- Check for invalid symbols
+    bad := select(symbols, (name, symb) -> not okay(name, symb));
+    if #bad > 0 then error(
+	"encountered symbol(s) that are not alphanumeric or have less than 2 characters:",
+	concatenate apply(bad, (name, symb) -> {
+		"\n\t",
+		-* TODO: symbolLocation name, ": here is the first use of ", *-
+		toString name}));
+
+    -------------------------------------------------------------------------------
+    -- Put the symbols into bins
+    cachedSymbols.Symbol   = first \ select(symbols,
+	(name, symb) ->isAlphaNumeric name);
+    cachedSymbols.Keyword  = first \ select(symbols, isKeyword);
+    cachedSymbols.Type     = first \ select(symbols, isType);
+    cachedSymbols.Function = first \ select(symbols, isFunction);
+    cachedSymbols.Constant = sort join(
+	first \ select(symbols, isConst),
+	{"Node", "Item", "Example", "CannedExample", "Pre", "Code", "Tree",
```

Comment:
Do you mind adding a comment close to this, and also somewhere in SimpleDoc, to keep these in sync?

Reply:
Is there a reason we hardcode these here instead of just exporting them from `SimpleDoc` and letting them get pickedup with all the other symbols?  (I suppose `Node` is a class, so it would get a different color -- is that the main reason?)

Reply:
I just pushed a commit exporting these docstring symbols from `SimpleDoc` so we can stop hard-coding them here.  I also renamed the `Node` class to `DocNode` so `Node` would still be the same color as the others.